### PR TITLE
TS-2150: Add Milestone log tags

### PR DESCRIPTION
**Pull request by fpesce**
- Add Milestone log tags
- Add Milestone diff log tags

### CODE REVIEW

**Comment by zwoop** [comment]
Please check with John (the current assignee of the TS-2150 Jira).

File: proxy/StatSystem.h [hunk 1]
```diff
@@ -49,69 +49,75 @@
...
-
-  // TODO: Should we instrument these at some point?
-  // ink_hrtime  cache_read_begin;
-  // ink_hrtime  cache_read_end;
-  // ink_hrtime  cache_write_begin;
-  // ink_hrtime  cache_write_end;
+  { // constructor to initialize array elements to zero in case user wants to
+    for (int i = 0; i < LAST_ENTRY; i++)
+      milestones[i] = 0;
+  }
```
Comment (SolidWallOfCode) [inline proxy/StatSystem.h]: `ink_zero(milestones);`

File: proxy/StatSystem.h [hunk 2]
```diff
@@ -49,69 +49,75 @@
...
+
+  void ms_set(Milestone ms, ink_hrtime val);
+  ink_hrtime ms_get(Milestone ms) const;
```
Comment (SolidWallOfCode) [inline proxy/StatSystem.h]: `operator[]` ?

File: proxy/http/HttpSM.cc [hunk 3]
```diff
@@ -6662,48 +6668,61 @@ HttpSM::update_stats()
...
+      "server_first_read: %.3f "
+      "server_read_header_done: %.3f "
+      "server_close: %.3f "
+      "ua_close: %.3f "
+      "sm_finish: %.3f "
+      "plugin_active: %.3f "
+      "plugin_total: %.3f",
+      sm_id, client_ip, ats_ip_port_host_order(&t_state.client_info.addr), url_string, status, unique_id_string, redirection_tries,
+      client_response_body_bytes, fd, t_state.client_info.state, t_state.server_info.state,
+      milestone_difference(milestones.ms_get(TransactionMilestones::SM_START), milestones.ms_get(TransactionMilestones::UA_BEGIN)),
```
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: Is this done enough to justify `TransactionMilestones::different(enum, enum)` ?

File: proxy/http/HttpTransact.cc [hunk 4]
```diff
@@ -8856,75 +8856,100 @@ HttpTransact::update_size_and_time_stats(State *s, ink_hrtime total_time, ink_hr
 
   // update milestones stats
   if (http_ua_begin_time_stat) {
-    HTTP_SUM_TRANS_STAT(http_ua_begin_time_stat, milestone_difference_msec(milestones.sm_start, milestones.ua_begin))
+    HTTP_SUM_TRANS_STAT(http_ua_begin_time_stat, milestone_difference_msec(milestones.ms_get(TransactionMilestones::SM_START),
+                                                                           milestones.ms_get(TransactionMilestones::UA_BEGIN)))
```
Comment (SolidWallOfCode) [inline proxy/http/HttpTransact.cc]: Here, at least, `milestone_difference_msec()` should take the enum values and do the `ms_get` internally. That would look a lot cleaner.

File: proxy/logging/LogField.cc [hunk 5]
```diff
@@ -212,6 +269,51 @@ LogField::~LogField()
...
+
+int
+LogField::milestones_from_m_name(const char *m_name, TransactionMilestones::Milestone *ms1, TransactionMilestones::Milestone *ms2)
+{
+  milestone_map::iterator it;
+  char *ms_name;
+  size_t ms_name_len, skip;
+
+  ms_name_len = strcspn(m_name, " -");
+  ms_name = strndup(m_name, ms_name_len);
```
Comment (SolidWallOfCode) [inline proxy/logging/LogField.cc]: It's very frowned upon to do allocations in the main path. The find method should be changed to accept a `char const*` and length and avoid the duplication / allocation here.

File: proxy/logging/LogField.cc [hunk 6]
```diff
@@ -212,6 +269,51 @@ LogField::~LogField()
...
+  it = m_milestone_map.find(ms_name);
+  free(ms_name);
+  if (it != m_milestone_map.end()) {
+    *ms1 = it->second;
+  } else {
+    for (it = m_milestone_map.begin(); it != m_milestone_map.end(); ++it)
```
Comment (SolidWallOfCode) [inline proxy/logging/LogField.cc]: What is this point of this loop, if it returns on the first iteration?

**Comment by SolidWallOfCode** [comment]
I think the use of `std::map` is problematic because of its secondary effects, most importantly having to duplicate a string just to do a look up is a serious flaw IMHO. You might look at using `ts::Buffer` or `ts::ConstBuffer` as the key type instead. Then you can use a length based comparison functor for the `std::map` comparator.

**Comment by bryancall** [comment]
We shouldn't be using std:map in the first place.  It would duplicate the string object, but not the string bytes.

**Comment by zwoop** [comment]
+1. No STL on the critical path.

**Comment by fpesce** [comment]
I've removed the allocation, std::map is used only once (as a singleton to turn string from conf into enum value). Is there any lookup table system/class already existing that you would me to use instead?

File: proxy/logging/LogField.cc [hunk 7]
```diff
@@ -132,6 +168,13 @@ LogField::LogField(const char *name, const char *symbol, Type type, MarshalFunc
 
   m_time_field = (strcmp(m_symbol, "cqts") == 0 || strcmp(m_symbol, "cqth") == 0 || strcmp(m_symbol, "cqtq") == 0 ||
                   strcmp(m_symbol, "cqtn") == 0 || strcmp(m_symbol, "cqtd") == 0 || strcmp(m_symbol, "cqtt") == 0);
+
+  if (m_milestone_map.empty()) {
```
Comment (SolidWallOfCode) [inline proxy/logging/LogField.cc]: This should really be in a local function or method, since it's used multiple times.

**Comment by SolidWallOfCode** [comment]
Looking much better. I think the `std::map` use is acceptable - it's not modified on the critical path and so there are no allocations which should be fine. Searching in `std::map` should be fast, especially for such a small set of strings. The initialization, though, shouldn't be cut and pasted so many places. Maybe that should be done in `Log::init_fields` as it is part of the field initialization.

My only other general comment is that sometimes `Milestone::diff` is used and sometimes the value are directly subtracted. That's a minor quibble, though. The real improvement there was from the millisecond computations which look a lot cleaner now.

**Comment by SolidWallOfCode** [comment]
Ready to ship. The only tweak I'd recommend now is to put a reference to `TSHttpTxnMilestoneGet` in the documentation update so there's a link to the list of milestones.

**Comment by bryancall** [comment]
I think it would be better at the creation of the LogField store the index into the milestone array instead of looking it up every time you log.  You can create a MilestoneLogFiled and have LogField as the base class if you don't want add a index to every LogField.

**Comment by SolidWallOfCode** [comment]
I discussed this with Bryan and Susan and we think the least worst option is to put two milestone indices in the `LogField` class. The code in `LogFormat.cc` that creates the `LogField` instances has no provisions for creating sublcasses, which IMHO is a problem but outside the scope of this bug. Therefore we recommend adding two milestone indices to `LogField` and storing the appropriate milestone indices there in the constructor (by doing the lookup in the map)  and then using those (instead of doing the lookup) during actual logging operations.

**Comment by fpesce** [comment]
Looks like a good least worst idea to me. I'll update the PR.

**Comment by shinrich** [comment]
The changes look good to me.  Please resolve the merge conflict, and I'll get this committed.

**Pull request closed without merging**